Experimental Features Menu (No Experimental Settings Yet)

jbienz opened the pull request:
Added new menu for Experimental Features under All Settings. The first time the user enters the menu they are presented with a dialog warning about features being a work in progress. This dialog can opted out for future runs, but even if the dialog is opted out the Sidebar still shows a warning while in the menu.

There are no experimental settings in this pull request, but my Whammy FX PR has been updated to show its settings under Experimental.

https://github.com/YARC-Official/YARG/pull/812

![O2HXaYhpXz](https://github.com/YARC-Official/YARG/assets/2062886/179eb1e2-af8d-4280-a1df-e188fa19ac13)

![image](https://github.com/YARC-Official/YARG/assets/2062886/321af2b9-2100-4c6c-bb16-5f2b3a6809eb)

![image](https://github.com/YARC-Official/YARG/assets/2062886/d23dc9f3-b7c1-4612-83a3-a3e801cda6a7)

RileyTheFox commented:
Before I test this I just want to mention that in future all PRs should target the `dev` branch. `master` is for stable releases and is merged to from `dev`, which is our 2 per day nightly release branch. I've changed this for you but in future try to remember!

jbienz commented:
> Before I test this I just want to mention that in future all PRs should target the `dev` branch. `master` is for stable releases and is merged to from `dev`, which is our 2 per day nightly release branch. I've changed this for you but in future try to remember!

I'm sorry Riley. I was aware of that requirement and I thought I had been careful to follow it. I will be even more vigilant next time.

EliteAsian123 commented on Assets/Script/Settings/Metadata/PreviewBuilders/ExperimentalPreviewBuilder.cs:
```diff
@@ -0,0 +1,42 @@
+        public async UniTask BuildPreviewUI(Transform uiContainer)
+        {
+            var go = Object.Instantiate(_experimentalPreviewUI, uiContainer);
+
+            // Enable and wait for layouts to rebuild
+            await UniTask.WaitForEndOfFrame(SettingsMenu.Instance);
+
+            // Skip the game object was somehow destroyed
+            if (go == null) return;
+
+            // No other configuration required
```
These statements aren't needed as there is nothing after them (only the instantiation is required).

jbienz replied:
I removed these lines of code and instead show the dialog which can be opted out.

EliteAsian123 commented on Assets/Script/Settings/SettingsManager.Settings.cs:
```diff
@@ -336,6 +336,8 @@ public void OpenExecutablePath()
 
             public ToggleSetting ShowAdvancedMusicLibraryOptions { get; } = new(false);
 
+            public ToggleSetting ShowExperimental { get; } = new(false, ShowExperimentalCallback);
+
```
Having a toggle to show/hide the settings might be a little overkill, as people may not know that the experimental settings exist. Showing the tab at all times with the warning should be good enough.

jbienz replied:
I'm fine with showing it all the time. @RileyTheFox seemed a little concerned about the support issues that could be raised, though that's why I added the big red text. @RileyTheFox what are your thoughts? Would you prefer it to stay hidden under a developer option or are you OK with it being visible all the time?

RileyTheFox replied:
I think settings being explicitly under an experimental tab should be enough to tell users that the features there are incomplete/prone to issues. I raised concern in regards to #812 as it would have originally been added to the audio settings, and I didn't want people thinking its a complete feature then bombarding us with help/bug reports when we know full well it has problems. 

For the experimental tab, we could maybe show a popup for the first time the user clicks on it explaining that everything inside it is work in progress/subject to change/will contain bugs. Thoughts on that? @EliteAsian123

EliteAsian123 replied:
Yeah, it can be one of those dialog boxes with the "don't show again" checkbox.

jbienz replied:
I made the menu visible by default but now show a dialog when the menu is first displayed. The dialog can be opted out similarly to the piracy dialog.

EliteAsian123 commented on Assets/Settings/Localization/Settings_en-US.asset:
```diff
@@ -1553,6 +1553,29 @@ MonoBehaviour:
+    m_Metadata:
+      m_Items: []
+  - m_Id: 168311354298281984
+    m_Localized: 'IMPORTANT
+
+
+      These settings are experimental and are not supported.'
```
Maybe change this to "these settings are experimental and could cause major problems if enabled."

jbienz replied:
I updated the text as suggested.

EliteAsian123 reviewed (changes requested):
Hi, thanks for the PR! I've submitted some minor feedback below.

The tab icon doesn't really artistically fit with the rest, however, I'll replace it later so don't worry about it.

jbienz commented:
Updated TabIcon to use Lucide icon. Made Experimental tab always visible under All Settings. Added dialog shown when entering Experimental Features. This dialog is shown once per launch and can also be marked to not show again. Added new localized strings for the dialog and title. Removed old localized  strings for the previous setting to show and hide the Experimental section.

EliteAsian123 commented:
This looks good to me for merging, however, there were just some localization changes pushed that will require some changes in this PR. After that, I'll merge! Thanks.

jbienz commented:
The localization assets appear to have been removed and I don't see any documentation on how strings should now be localized in your new system. @EliteAsian123 can you please make the required changes, or please tell me the necessary steps to use your new localization system?
